File: 算法_js数据结构+python/js/数据结构与类型/dbLinkedList.js
function dbLinkedList(){
    let Node = function(element){
        this.element = element;
        this.next = null;
        this.pre = null;
    }
    let head = null,
        length = 0,
        tail = null;

    
    this.insert = function(pos,element){
        if(pos > -1 && pos <= length){
            let node = new Node(element),
                current = head,
                prev = null,
                index = 0;
            if(pos===0){
                if(!head){
                    head = node;
                    tail = node;
                }else{
                    node.next = current;
                    current.pre = node;
                    head = node;
                }
            }else if(pos===length){
                current = tail;
                current.next = node;
                node.pre = current;
                tail = node;
            }else{
                while(index ++ < pos){
                    prev = current;
                    current = current.next;
                }
                node.next = current;
                prev.next = node;

                current.pre = node;
                node.pre = prev;
            }
            length ++ ;
            return true;
        }else{
            return false;
        }
    };
    this.removeAt = function(pos){
        if (pos > -1 && pos < length){
            let current = head,
                index = 0,
                prev = null;
            if(pos === 0){
                head = current.next;
                if(length === 1){
                    tail = null;
                }else{
                    head.prev = null;
                }
            }else if (pos === length - 1){
                current = tail;
                tail = current.pre;
                tail.next = null;
            }else{
                while (index ++ < pos){
                    prev = current;
                    current = current.next; 
                }
                
                prev.next = current.next;
                current.next.pre = prev
            }
            length --;
            return current.element;
        }
        else{
            return null;
        }
    };

    
}

File: 算法_js数据结构+python/python/数据结构与类型/dbLinkedList.py
class Node():
    def __init__(self,data):
        self.data = data
        self.next = None
        self.prev = None

class dbLinkedList():
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0
    def insert(self,pos,item):
        node = Node(item)
        current = self.head
        previous = None
        index = 0
        if pos > -1 and pos <= self.length:
            if pos == 0:
                if self.head==None:
                    self.head = node
                    self.tail = node
                else:
                    node.next = current
                    current.prev = node
                    self.head = node
            elif pos == self.length:
                current = self.tail
                current.next = node
                node.prev = current
                self.tail = node
            else:
                while index < pos:
                    previous = current
                    current = current.next
                    index += 1
                previous.next = node
                node.next = current
                current.prev = node
                node.prev = previous
            self.length += 1
        else:
            return False
    def removeAt(self,pos):
        current = self.head
        previous = None
        index = 0
        if pos > -1 and pos < self.length:
            if pos == 0:
                self.head = current.next
                if self.length == 1:
                    self.tail = None
                else:
                    self.head.prev = None
            elif pos == self.length - 1:
                current = self.tail
                self.tail = current.prev
                self.tail.next = None
            else:
                while index < pos:
                    previous = current
                    current = current.next
                previous.next = current.next
                current.next.prev = previous
            self.length -= 1
            return current.data
        else:
            return None


    def size(self):
        return self.length

db = dbLinkedList()
db.insert(0,1)
db.insert(0,'mk')
db.insert(0,'lxw')
db.removeAt(0)
print(db.size())
